src/expression/ExpressionVisitor.cpp

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Team Dissolve and contributors
+
+#include "expression/ExpressionVisitor.h"
+#include "data/ff.h"
+#include "expression/ExpressionErrorListeners.h"
+#include "expression/binary.h"
+#include "expression/functionNEW.h"
+#include "expression/number.h"
+#include "expression/reference.h"
+#include "expression/unary.h"
+#include "expression/variable.h"
+#include "templates/optionalref.h"
+
+/*
+ * Creation Entry Point
+ */
+
+// Return the topmost context in the stack
+std::shared_ptr<ExpressionNodeNEW> ExpressionVisitor::currentContext() const
+{
+    assert(contextStack_.size() != 0);
+
+    return contextStack_.back();
+}
+
+// Construct description within supplied object, from given tree
+void ExpressionVisitor::create(ExpressionNEW &expr, ExpressionParser::ExpressionContext *tree,
+                               RefList<ExpressionVariable> externalVariables)
+{
+    expression_ = &expr;
+    externalVariables_ = externalVariables;
+
+    // Add the definition's root node as the topmost context in the stack
+    contextStack_.push_back(expression_->rootNode());
+
+    // Traverse the AST
+    visitChildren(tree);
+}
+
+/*
+ * Visitor Overrides
+ */
+
+antlrcpp::Any ExpressionVisitor::visitParentheses(ExpressionParser::ParenthesesContext *ctx)
+{
+    auto node = std::make_shared<ExpressionRootNode>();
+
+    currentContext()->addChild(node);
+
+    contextStack_.push_back(node);
+
+    auto result = visitChildren(ctx);
+
+    contextStack_.pop_back();
+
+    return result;
+}
+
+antlrcpp::Any ExpressionVisitor::visitUnaryMinus(ExpressionParser::UnaryMinusContext *ctx)
+{
+    auto node = std::make_shared<ExpressionUnaryOperatorNode>(ExpressionUnaryOperatorNode::OperatorNegate);
+
+    currentContext()->addChild(node);
+
+    contextStack_.push_back(node);
+
+    auto result = visitChildren(ctx);
+
+    contextStack_.pop_back();
+
+    return result;
+}
+
+antlrcpp::Any ExpressionVisitor::visitPower(ExpressionParser::PowerContext *ctx)
+{
+    auto node = std::make_shared<ExpressionBinaryOperatorNode>(ExpressionBinaryOperatorNode::OperatorPow);
+
+    currentContext()->addChild(node);
+
+    contextStack_.push_back(node);
+
+    auto result = visitChildren(ctx);
+
+    contextStack_.pop_back();
+
+    return result;
+}
```

Comment:
It's largely a mater of style, but I would invert the logic here.

```suggestion
antlrcpp::Any ExpressionVisitor::visitPower(ExpressionParser::PowerContext *ctx)
{
    return ExpressionBinaryOperatorNode(visit(ctx->left(), ctx->right());
}
```

This would, of course, involve labelling the left and right terms in the grammar file.  This has the advantage of letting the compiler handle the stack for us, plus everything remains immutable values.